### DISCUSSION TITLE
Scale Mesa to Model Millions of Agents

### DISCUSSION THREAD

**Discussion by Srj** [mesa/mesa#1033] [Ideas]
Is there a way i can take advantages of my multiple core to parallelize computation of mesa. For example, In Wolf-Sheep predator model if i use 100k wolf and 300k Sheep the model runs very very slowly. I guess there is a way to parallalize the operation?

**Comment by tpike3** [comment]
I have not found a good way to parallalize a single model. As Python can really only do multiprocessing due to the Global interpreter lock (GIL) I tried using message passing to split a model spatially and pass as agent crossed the boundaries. But in general message passing is too slow so yo don't gain a lot of efficiencies. 

I think the more appropriate way would be to convert a model to arrays but I have not attempted that yet. Defnieitely interested in more discussion on the topic.

**Reply by Srj** [reply]
How about using concurrent Process pool executor? https://docs.python.org/3/library/concurrent.futures.html

**Reply by tpike3** [reply]
@Srj I personally have not tried that. But, if you can find a way I am all for it. 

The challenge we keep running into is how to do it so it is generic for any model vs the optimization of one model.

**Reply by rht** [reply]
`concurrent.futures` is just another different API to schedule executions, and so using it won't necessarily result in a faster code than using `multiprocessing`.

**Reply by jackiekazil** [reply]
I am late to the party. This is all very interesting. Has anyone explored any of these?

**Reply by rht** [reply]
@DavoudTaghawiNejad has recently written a model using Numpy arrays to store agents states.

**Comment by TabeaSonnenschein** [comment]
Hi everyone, interesting discussion! Any update on an easy implementation of multiprocessing or concurrent.future for the agent's steps in the mesa model? Maybe in 2 years the outlook has changed? Thanks :).

**Reply by rht** [reply]
Thank you for posting here. It further signals about the demand for a scalable Mesa. The building blocks for this to happen already exist, and just need coordination for it to materialize.

**Reply by Corvince** [reply]
Yes, thank you for posting. Maybe you could write a bit more about your use case? We are indeed collecting use cases that require a scalable mesa. 

Unfortunately, there can't be an easy implementation of multithreaded agents. The advantage of agent based models is that agents interact with each other and as soon as you have interaction, multithreading becomes a hard problem. 
Luckily there are different approaches to speed things up, so hearing about your use case will really help us to provide an easy way out.

**Reply by TabeaSonnenschein** [reply]
Thank you so much @rht  and @Corvince for your quick replies. 

My use case: The simulation models that I am building and researching integrate behavioral models of urban residents (e.g. mobility, activity and consumption choices) and dynamic physical models of environmental stressors (e.g. air pollution, noise, heat), which result in an interaction that approximates personal exposure. Consequently, the accumulated health impacts for different intervention scenarios and population groups can be predicted using exposure-response functions. 

So far I am not super interested in the social interaction between agents and more in the collective impact that agents have on the environment (e.g. air pollution) and the impact the environment has on the agents in return (e.g. health). Since agents do not directly interact with each other but only influence each other through my environmental models, I can parallelize the agents step function. Before switching to mesa because I find python handy, I have used GAMA, which also allows parallelizing the agents behavior.

Below some first steps I attemped (1 day tryout) using multiprocessing's Pool:

1. For the initialization of the agents, I did:
      
       with multiprocessing.Pool() as pool:
                  agents = pool.starmap( Humans, [(list(random_subset.iloc[i]), self) for i in range(self.nb_humans)])
      
 However, parallelization turned out slower than sequential initialization due to overhead and my inadequant management of shared resources. If you know how to improve this effort, I would highly appreciate your help!

2. For parallelizing the agent's step function, I tried generating a new scheduler: 

          class ParallelScheduler(BaseScheduler):
              def __init__(self, model):
                  super().__init__(model)
             
              def step(self):
                  # Shuffle the agents to avoid any potential bias in the order of execution
                  self.model.random.shuffle(self.agents)
                  # Use the multiprocessing pool to execute the agents' steps in parallel
                  with multiprocessing.Pool() as self.pool:
                    self.pool.map(self.step_agent, self.agents)
          
              @staticmethod
              def step_agent(agent):
                  agent.step()

However, this turned out to reinitiate the model for each core again and again, which is useless and unefficient. I guess I need to optimize the shared resource management. Perhaps you have an idea? I have tons of spatial data in the Model that the agents interact with.


3. Finally, I am trying to translate the model into Cython, but that is also a lot of learning for me and I am struggling a bit to find the time to do all of this next to my other model development tasks. [#1610](https://github.com/projectmesa/mesa/discussions/1610). Maybe I can actually forget it, because I use spatial datatypes and geopandas, ect? Or could it still work?

Sorry for the long essay :D. It is a nitty griddy topic isn't it? Thanks for your response and if you feel like answering, it would be highly appreciated.
Cheers,
Tabea

**Reply by Corvince** [reply]
Just quickly two good news for you. First I think you don't need to worry about multiprocessing. Since you want to evaluate different scenarios you can use multiprocessing then. That is you leave each model run single processed, but run multiple simulations in parallel. Have a look at mesas `batch_run` function. That should be the fastest option.

Second, you can try out [https://github.com/VU-IVM/honeybees](). I haven't used it personally, but it should fit your use case where you only have interactions with the environment. 

In any ways please keep us updated

**Reply by Corvince** [reply]
> Finally, I am trying to translate the model into Cython, but that is also a lot of learning for me and I am struggling a bit to find the time to do all of this next to my other model development tasks. #1610. Maybe I can actually forget it, because I use spatial datatypes and geopandas, ect? Or could it still work?

Are you using mesa-geo or how do you interact with your geospatial data?

**Reply by rht** [reply]
In addition to honeybees, there is also Sampy (discussion at https://github.com/projectmesa/mesa/discussions/1561). To contextualize, you represent individual agents as rows in an array or DF, and it could run on a laptop hardware. You could also use `cupy` to run on GPU.

**Reply by TabeaSonnenschein** [reply]
Hi, again thanks for the quick response. Unfortunately the batch run function is not what I need. I need to run sufficient agents to have a traffic assignment model (e.g. 10% of the population of Amsterdam), because traffic causes air pollution, which in return impacts agents. 

Cool, I will look at honeybees and Sampy. I am also looking at Agents.jl right now, because it is much faster. 

I do like python though. So if you know of an easy template to implement multiprocessing with efficient data management in mesa, let me know.  Otherwise, SamPy could be interesting indeed. Will have a look at the applicability of cupy too.

About the spatial computation, I simply use a continousSpace and a consistent CRS, so if I want agents to move I assign the coordinates  to their pos. Their have residences and workplaces and other POI or spatial data defined destinations. I used the Open Source Routing machine and lots of spatial data to have a realistic interaction with the environment. So far I have not visualized the mesa ABM. So, no I do not use mesa-geo.

**Reply by Srj** [reply]
I posted here a few years back when I was actually simulating a COVID-19 model for 200k who are interacting with each other and spreading the disease.

I tried different parallel processing libraries/frameworks such as multiprocessing, ray, and many more I don't quite remember. They didn't quite improve my speed of processing agents but one pitfalls I saw using these types of parallelizing is that the states of the agents are updated in place. So, Say you have 10 agents and update them using 10 threads. Now say Agent 5 is updated early and then the thread of Agent 1 is started. Then Agent 1 will use the updated information from future for Agent 5 and that will make your simulation wrong. So, based on the race conditions of the threads the results will highly vary. So, I wanted to give you this warning. 
To overcome this, I thought I might make 2 copies of each agent and when I am updating I will update the second matrix using the state information from the first matrix and in the end I will replace the first matrix with the second. But I really got busy with other projects and never implemented the idea. Maybe I would like to know if you can make mesa scalable. 

Also there is a framework name GAMA (https://gama-platform.org/) that I heard of that it is highly scalable and can even use GPU and very efficient with spatial data. I never used it though but you can give a try.

**Reply by TabeaSonnenschein** [reply]
Hi @Srj , I have my model implemented in GAMA but it was too slow. It seems that mesa does not really help with that. It is better at certain things, but the lack of multiprocessing is a bummer indeed. Thanks for sharing your idea. I still have to make a decision on how I will proceed.

**Reply by rht** [reply]
I second with @Srj's statement that you will run into non-deterministic execution order. I had encountered this in the past with @DavoudTaghawiNejad and abcEconomics, and we solved it with splitting the model step into sending messages among the agents to their inboxes, and actually doing the actions. So the decision making and the action are separated. That said, @TabeaSonnenschein, I recommend you to investigate a bit more on why your GAMA code is slow. GAMA seems worth it to me.

**Reply by TabeaSonnenschein** [reply]
Hi thanks!
Well it is slow because of the spatial processing. Additionally, I have difficulty loading my pmml decision tree model, or perform my cellular automata dispersion model in a fast and efficient way. I will see if I can make it work faster in GAMA, but I feel constraint by the GAML language.

**Reply by TabeaSonnenschein** [reply]
@Srj , thanks for explaining. Fortunately, the agents do not depend on each other during a step. They just have a collective impact on the environment. Why do you think the parallelization did not improve the speed? Was that the shared resources and overhead issue?

@Srj  and @rht, if you have example code on how you implemented the parallelization and/or you represent individual agents as rows in an array or DF, that would help me immensely!

**Reply by rht** [reply]
Can you post a minimal reproducible demo of your model, with unnecessary details simplified?

**Reply by TabeaSonnenschein** [reply]
I excluded some major chucks but left a sceleton. Does that work? Without the data, you cannot run it anyway, I guess. Thanks for your time. I truly appreciate! :-)

** The model is still under development

**Reply by rht** [reply]
Can you remove small unnecessary details. Just keep the 0th or 1st order effect. This is so that it is easy to explore, before applying it on the full version. Have you tried using https://github.com/pyutils/line_profiler to measure which part is the slowest? Also, linking to a file posted on a GitHub repo is better.

**Reply by TabeaSonnenschein** [reply]
@everyone . Thanks for thinking along with me. It really helped me. Yesterday, I found out that the slowness of my ABM parallelization efforts on Windows was caused by the fact that Windows uses "spawning" to start new workers, which creates new versions of all the variables for each worker. On the other hand Linux allows for a technique called "forking", in which the child process inherits all the same variables in the same state as they were in the parent. Each child process then continues independently from the forking point. This reduces the overhead of dataintensive tasks to the transfer of variables that are actually changing.

Using that new knowledge I set up my WSL Ubuntu and tried out the concurrent.futures (cf) ProcessPoolExecutor. In other words, the good news is that I managed to parallelize the ABM and gain significant speed improvements. Yey! To make things explicit for future computational venturers:

Creating agents during init. Map the Class function (in my case Humans) to the iteratables. Best to load all data that is needed in the init function of the agents before so that it can be accessed with the iterable. You can add as many iterables as you want to the map functions, but try keep it as short as possible.
```
        with cf.ProcessPoolExecutor() as executor:
            agents = executor.map(Humans, range(self.nb_humans), it.repeat(self, self.nb_humans)) 
        for agent in agents:
            self.continoussp.place_agent(agent, agent.Residence)
            self.schedule.add(agent)

```
Now to the interesting part: Parallelizing the agent's step function. For that you replace the "self.schedule.step()" of the model.step function with the code below. Interestingly, the ThreadPoolExecutor automatically saved all updated agent variables when simply executing the Humans.step function. However, it does not yet use all CPU power. When exchanging the ThreadPoolExecutor with the ProcessPoolExecutor all CPU power is used, however the variables are not yet saved. I am still figuring out how that works.
```
        with cf.ThreadPoolExecutor() as executor:
            executor.map(Humans.step, self.schedule.agents, it.repeat(self.current_datetime, self.nb_humans) )    
```
Anyhow, during my journey (into the unknown), I noticed that you can very easily also parallelize only subfunctions of the agents. This allows bypassing the interaction problem that you all keep mentioning, and which I agree with. E.g. So you can parallelize all functions that are independent of agent interaction with something similar to the code below.
```
        with cf.ProcessPoolExecutor() as executor:
           for agent in self.schedule.agents:     
               if agent.traveldecision == 1:
                 agent.RouteVars, agent.OrigVars, agent.DestVars = executor.submit(PerceiveEnvironment,agent.route_eucl_line, agent.pos, agent.destination_activity, self.EnvBehavDeterms).result()                
```        

Then you can call functions that are depedent on interaction in the agent.step function using "self.schedule.step()" before or after. Or/And you can also iterate through the agents to call that function before or after the parallelized function call, e.g.

```    
   for agent in self.schedule.agents:     
            agent.YourFunction-()

```    

Hope this helps others!  @rht , yes indeed, profilers are useful for optimization too!

**Comment by Srj** [comment]
> Why do you think the parallelization did not improve the speed? Was that the shared resources and overhead issue?

@TabeaSonnenschein I think in my case it didn't improve because each agent had to interact with all other agents. As you said this is not the case for you, it might improve the speed in your case. It would be great if you give an update here when/if you implement it.

**Reply by TabeaSonnenschein** [reply]
@Srj yes that seems like an exponential problem if they have to interact with ALL other agents. My issue was the resource management, since I have a lot of data in my model and the transfer to all other workers is slow. However, I found a solution based on Linux fork technique: (see response here https://github.com/projectmesa/mesa/discussions/1033#discussioncomment-6625908)
